### PR TITLE
feat: conversation message sent/delivered/read status and reactions DB infra, unit tests

### PR DESCRIPTION
**Pull request by ice-damocles**
## Description
Adds:
- `isSent`, `isReceived`, `isRead` column for conversation messages table
- Mark conversation message as `isSent` method for DB notifier (after "OK" response from relay)
- Mark conversation message as `isReceived` or `isRead` (all previous received messages)
- New `conversationReactionsTable` in DB
- Insert kind 7 "received" and "read" content event messages to DB and update `conversationReactionsTable`
- Get reactions for conversation message

## Type of Change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [x] Refactoring
- [ ] Documentation
- [ ] Chore

### CODE REVIEW

**Review by ice-kreios — DISMISSED** [review]

File: lib/app/features/chat/model/entities/reaction_data.c.dart [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+import 'package:collection/collection.dart';
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:ion/app/exceptions/exceptions.dart';
+import 'package:ion/app/features/chat/model/entities/private_direct_message_data.c.dart';
+import 'package:nostr_dart/nostr_dart.dart';
+
+part 'reaction_data.c.freezed.dart';
+
+@immutable
```
Comment (ice-alcides) [inline lib/app/features/chat/model/entities/reaction_data.c.dart:11]: `@immutable` is redundant here

File: lib/app/features/chat/model/entities/reaction_data.c.dart [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+@Freezed(equal: false)
+class ReactionEntity with _$ReactionEntity {
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/features/chat/model/entities/reaction_data.c.dart:13]: Initially I used it but it doesn't fit the conversation kind 7 as doesn't have `masterPubkey` and is not encrypted etc.

File: lib/app/features/chat/model/entities/private_direct_message_data.c.dart [hunk 3]
```diff
@@ -67,6 +68,7 @@ class PrivateDirectMessageData with _$PrivateDirectMessageData, EntityMediaDataM
   const factory PrivateDirectMessageData({
     required List<TextMatch> content,
     required Map<String, MediaAttachment> media,
+    @Default([]) List<ReactionEntity> reactions,
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/features/chat/model/entities/private_direct_message_data.c.dart:71]: Actually it is not even entity, I have to rename it

File: lib/app/features/chat/model/entities/reaction_data.c.dart (hunk 1 above)
Comment (ice-damocles) [reply lib/app/features/chat/model/entities/reaction_data.c.dart:11]: Needed for == and hashtag override
image

File: lib/app/services/database/ion_database.c.dart [hunk 4]
```diff
@@ -220,11 +366,26 @@ class ConversationMessagesTable extends Table {
...
+  BoolColumn get isReceived => boolean().withDefault(const Constant(false))();
+  BoolColumn get isRead => boolean().withDefault(const Constant(false))();
 
   @override
   Set<Column<Object>> get primaryKey => {eventMessageId};
 }
 
+// Table for conversation reactions (is automatically updated when
+// [EventMessagesTable] is updated with new records)
+class ConversationReactionsTable extends Table {
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/services/database/ion_database.c.dart:379]: Sounds better together with previous database table convention
image

File: lib/app/services/database/ion_database.c.dart [hunk 5]
```diff
@@ -220,11 +366,26 @@ class ConversationMessagesTable extends Table {
...
+  TextColumn get content => text()();
+  TextColumn get id => text()();
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/services/database/ion_database.c.dart:381]: Okay

File: lib/app/features/chat/model/entities/private_direct_message_data.c.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/features/chat/model/entities/private_direct_message_data.c.dart:71]: Removed

File: lib/app/services/database/ion_database.c.dart [hunk 6]
```diff
@@ -194,6 +314,32 @@ class IONDatabase extends _$IONDatabase {
...
+
+    final reactions = reactionsEventMessages
+        .map(
+          (reactionEventMessageData) => ReactionEntity.fromEventMessage(
+            reactionEventMessageData.toEventMessage(),
+          ),
+        )
+        .toList();
+
+    return privateDirectMessageEntity.copyWith.data(reactions: reactions);
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/services/database/ion_database.c.dart:341]: Okay

File: lib/app/services/database/ion_database.c.dart [hunk 7]
```diff
@@ -220,11 +366,26 @@ class ConversationMessagesTable extends Table {
   DateTimeColumn get createdAt => dateTime()();
   TextColumn get subject => text().nullable()();
   BoolColumn get isDeleted => boolean().withDefault(const Constant(false))();
+  BoolColumn get isSent => boolean().withDefault(const Constant(false))();
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/services/database/ion_database.c.dart:369]: Done

File: lib/app/features/chat/model/entities/reaction_data.c.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/features/chat/model/entities/reaction_data.c.dart:13]: Renamed

**Review by ice-alcides — APPROVED** [review]

**Review by ice-tychon — APPROVED** [review]

**Pull request merged**